content/blog/2025-04-27-valkey-modules-rust-sdk-updates.md

```diff
@@ -0,0 +1,272 @@
+        KeyChangeSubevent::Expired => {}
+    }
+}
+```
+
+## Custom ACL categories support
+
+Valkey 8 introduced support for custom ACL categories.  To implement that we need to enable `required-features = ["min-valkey-compatibility-version-8-0"]` in `Cargo.toml` and register new categories in `valkey_module!` macro.  Then we can restrict our custom commands to these ACL categories.
+
+```rust
+valkey_module! {
+    ...
+    acl_categories: [
+        "acl_one",
+        "acl_two"
+    ]
+    commands: [
+        ["cmd1", cmd1, "write",  0, 0, 0, "acl_one"],
+        ["cmd2", cmd2, "", 0, 0, 0, "acl_one acl_two"],
+    ]
+```
+
+## Validating / Rejecting CONFIG SET
+
+SDK now supports specifying optional callback functions to validate or reject custom configuations.  This is available for String, i64, bool and enum configs.  Here is an example of validation for i64 custom config.  
+
+```rust
+lazy_static! {
+    static ref CONFIG_I64: ValkeyGILGuard<i64> = ValkeyGILGuard::default();
+}
+fn on_i64_config_set<G, T: ConfigurationValue<i64>>(
+    config_ctx: &ConfigurationContext,
+    _name: &str,
+    val: &'static T,
+) -> Result<(), ValkeyError> {
+    if val.get(config_ctx) == custom_logic_here {
+        log_notice("log message here");
+        Err(ValkeyError::Str("error message here "))
+    } else {
+        Ok(())
+    }
+}
+valkey_module! {
+    configurations: [
+        i64: [
+            ["my_i64", &*CONFIG_I64, 10, 0, 1000, ConfigurationFlags::DEFAULT, Some(Box::new(on_configuration_changed)), Some(Box::new(on_i64_config_set::<ValkeyString, ValkeyGILGuard<i64>>))],
+        ],
+        ...
+    ]
+}
+```
+
+## Defrag
+
+There is a new `Defrag` struct abstracting away the raw C FFI calls to implement defrag for custom data types.
+
+```rust
+static MY_VALKEY_TYPE: ValkeyType = ValkeyType::new(
+    "mytype123",
+    0,
+    raw::RedisModuleTypeMethods {
+        ...
+        defrag: Some(defrag),
+    },
+);
+unsafe extern "C" fn defrag(
+    defrag_ctx: *mut raw::RedisModuleDefragCtx,
+    _from_key: *mut RedisModuleString,
+    value: *mut *mut c_void,
+) -> i32 {
+    let defrag = Defrag::new(defrag_ctx);
+    ...
+    0
+}
+valkey_module! {
+    ...
+    data_types: [
+        MY_VALKEY_TYPE,
+    ],
+    ...
+}
+```
+
+## Redis support
+
+There is a new feature flag in case module needs to run on Valkey and Redis.  Specify `use-redismodule-api` so that module used RedisModule API Initialization for backwards compatibility.
+
+```rust
+[features]
+use-redismodule-api = ["valkey-module/use-redismodule-api"]
+default = []
+
+cargo build --release --features use-redismodule-api
+```
+
+## Unit tests and memory allocation
+
+This feature flag supports running unit tests outside of Valkey or Redis.  As the result it cannot use Vakey Allocator and instead relies on the System Allocator instead.  The core logis is present in `alloc.rs` but developers only need to specify this in the module Cargo.toml.
```

Comment:
Cargo.toml in back ticks to make it inline code

Reply:
updated